File: [23] Merge k Sorted Lists/23.merge-k-sorted-lists.cpp
/*
 * @lc app=leetcode id=23 lang=cpp
 *
 * [23] Merge k Sorted Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
// #include <vector>
// #include <queue>
// using namespace std;

// struct ListNode {
//     int val;
//     ListNode *next;
//     ListNode() : val(0), next(nullptr) {}
//     ListNode(int x) : val(x), next(nullptr) {}
//     ListNode(int x, ListNode *next) : val(x), next(next) {}
// };

class cmp {
public:
    bool operator() (ListNode *a, ListNode *b) {
        return a->val > b->val;
    }
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        // Compare every k nodes (head of every linked list) and get the node with the smallest value.
        // Optimize it by priority queue
        // Time complexity O(N * log k)
        // N is the total number of nodes in the result linked list
        if (lists.size() == 0) return nullptr;
        if (lists.size() == 1) return lists[0];
        priority_queue<ListNode*, vector<ListNode*>, cmp> q;
        for (ListNode* l: lists) {
            if (l) q.push(l);
        }
        ListNode dummy;
        ListNode *current = &dummy;
        while (!q.empty()) {
            current->next = q.top();
            q.pop();
            current = current->next;
            if (current->next) q.push(current->next);
        }
        return dummy.next;
        // without priority queue:
        // ListNode dummy;
        // ListNode *current = &dummy;
        // while (true) {
        //     int min_idx = 0;
        //     int minimum = INT_MAX;
        //     bool isBreak = true;
        //     for (int i = 0; i < lists.size(); i++) {
        //         if (lists[i] != nullptr) {
        //             isBreak = false;
        //             if (lists[i]->val < minimum) {
        //                 minimum = lists[i]->val;
        //                 min_idx = i;
        //             }
        //         }
        //     }
        //     if (isBreak) break;
        //     current->next = lists[min_idx];
        //     current = current->next;
        //     lists[min_idx] = lists[min_idx]->next;
        // }
        // return dummy.next;
    }
};
// @lc code=end

File: [23] Merge k Sorted Lists/23.merge-k-sorted-lists.java
/*
 * @lc app=leetcode id=23 lang=java
 *
 * [23] Merge k Sorted Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        // Apply Divide and Conquer
        // time complexity O(N * log k)
        // N is total number of nodes in the result linked list
        int k = lists.length;
        if (k == 0) return null;
        if (k == 1) return lists[0];
        int interval = 1;
        while (interval < lists.length) {
            System.out.println(lists.length);
            for (int i = 0; i + interval < lists.length; i = i+interval*2) {
                lists[i]=mergeTwoLists(lists[i],lists[i+interval]);            
            }
            interval *= 2;
        }

        return lists[0];
    }

    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        // Time complexity O(n1 + n2)
        // n1, n2 is the number of nodes for the two linked lists
        ListNode dummyhead = new ListNode(0);
        ListNode current = dummyhead;
        while (l1 != null && l2 != null) {
            int val1 = l1.val;
            int val2 = l2.val;
            current.next = (val1<val2)? l1 : l2;
            current = current.next;
            l1 = (val1<val2)? l1.next: l1;
            l2 = (val1<val2)? l2:l2.next;
        }
        current.next = (l1 != null)? l1: l2;
        return dummyhead.next;
    }
}
// @lc code=end

